Fix many bugs in the stats reporter

godlygeek opened the pull request:
Remove the `--include-all-allocations` / `-a` option for the `stats`
reporter and have it process all allocations by default. Its default
behavior of only processing allocations that made up the high water mark
of the application's memory usage was confusing and misleading.

Previously when `-a` was provided, deallocations with `free` and
`munmap` were included in the reported count of "Total allocations".
This has been fixed.

Similarly, when `-a` was provided, memory freed by `munmap` (though not
`free`) was incorrectly being added to the count of total memory
allocated. This has been fixed.

The histogram has been fixed to work on the actual sizes of all
allocations. Previously it only saw the sizes after a rollup by stack
had already been performed, and it was therefore binning allocation
sizes that had already been summed.

Fix the two "largest allocating locations" sections to actually
aggregate by location. Previously they were aggregating by distinct
stack, so if the same line of code was reached by two different paths,
it would be counted separately instead of together.

Exclude `PYMALLOC_FREE` from the allocator type distribution (other
deallocators were already being ignored, but this recently added one was
missed).

Additionally, this fixes the bug introduced in #134 where `memray stats -a` was incorrectly handling any pointer that was reused after being reallocated. I haven't included a changelog entry for that one since it was never included in any released version.

Closes: #85 
Closes: #134

pablogsal commented on news/136.bugfix.3.rst:
```diff
@@ -0,0 +1 @@
+Fix the ``memray stats`` histogram to be based on the actual sizes of all allocations. Previously it only saw the sizes after a rollup by stack had already been performed, and it was therefore binning allocation sizes that had already been summed.
```
```suggestion
Fix the ``memray stats`` histogram to be based on the actual sizes of all allocations. Previously it only saw the sizes after a rollup by stack had already been performed, and it was, therefore, binning allocation sizes that had already been summed.
```

pablogsal commented on news/136.bugfix.4.rst:
```diff
@@ -0,0 +1 @@
+Fix the two "largest allocating locations" sections in the ``memray stats`` report to actually aggregate by location. Previously they were aggregating by distinct stack, so if the same line of code was reached by two different paths, it would be counted separately instead of together.
```
```suggestion
Fix the two "largest allocating locations" sections in the ``memray stats`` report to actually aggregate by location. Previously they were aggregating by distinct stacks, so if two different paths reached the same line of code, it would be counted separately instead of together.
```

pablogsal commented on src/memray/reporters/stats.py:
```diff
@@ -202,21 +131,25 @@ def render(
+    def _get_allocator_type_distribution(self) -> Iterator[str]:
+        for allocator_name, count in sorted(
+            self._stats.allocation_count_by_allocator.items(),
+            key=lambda item: item[1],
```
Nit: you can use `operator.itemgetter` for better performance

godlygeek replied:
This dict contains one entry per allocator, and we've got less than 16 of them. I'm gonna keep the more readable, less performant one, simply because this dict is too small to care.

pablogsal commented on src/memray/reporters/stats.py:
```diff
@@ -1,98 +1,41 @@
-    dist = Counter(min((x - low) // step, bins - 1) for x in it)
+    dist: Dict[int, int] = Counter()
+    for size, count in data.items():
+        if size:
+            bucket = min(int((math.log(size) - low) // step), bins - 1)
+        else:
+            bucket = 0
```
Nit:

```suggestion
        bucket = min(int((math.log(size) - low) // step), bins - 1) if size else 0
```

pablogsal reviewed (changes requested):
We need to update the docs if we are removing the `-a` parameter and likely the text needs some tweaking as the semantics  are changing a bit.

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+        while True:
+            PyErr_CheckSignals()
+            ret = reader.nextRecord()
+            if ret == RecordResult.RecordResultAllocationRecord:
```
RecordResultRecordRecordRecordResultRecordEhIsARecordDontBeMistakenIsARecord

pablogsal replied:
For the curious observer: I named this 😈

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+    #       because Cython's iterator handling is not const-correct.
+
+    # Convert allocation counts by allocator/by size to Python dicts.
+    allocation_count_by_allocator = {
```
This is quite more expensive that it needs to be. This is creating an `AllocatorType` key and hashing it over for insertion. `AllocatorType` are enums in Python and those are quite expensive to build. It may be much faster to take the dictionary as it is (Cython automatically transforms C++ maps into dictionaries) and then create other where the keys are transformed.

godlygeek replied:
Discussed offline: this dict contains each allocator only once, so the cost of constructing this dict is absolutely trivial compared to all the work we just did reading the file.

The cost of hashing the `AllocatorType`, along with the cost of `+=` on a Python `int`, are the reason that I chose to build this up as a `unordered_map<int, size_t>` and only convert it to a Python `dict[AllocatorType, int]` at the very end.

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+            aggregator.allocationCountByAllocator()
+        )
+    }
+    allocation_count_by_size = {
+        size_and_count.first: size_and_count.second
+        for size_and_count in <unordered_map[size_t, uint64_t]&>(
+            aggregator.allocationCountBySize()
+        )
+    }
```
```suggestion
    allocation_count_by_size = aggregator.allocationCountBySize()
```

pablogsal replied:
Same as before, Cython should be able to return this automatically, no?

godlygeek replied:
I couldn't find the right magical incantation to make that happen, but Pablo and I played with it offline and got this working.

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+        (loc, size_and_count[0])
+        for loc, size_and_count in size_and_count_by_location.items()
+    ]
+    top_locations_by_size.sort(key=lambda pair: pair[1], reverse=True)
```
Nit: Using `operator.itemgetter` may be better than the lambda

godlygeek replied:
I rewrote this to use `heapq.nlargest` and avoided the need for `key` entirely.

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+        (loc, size_and_count[1])
+        for loc, size_and_count in size_and_count_by_location.items()
+    ]
+    top_locations_by_count.sort(key=lambda pair: pair[1], reverse=True)
```
Nit: Using `operator.itemgetter` may be better than the lambda

godlygeek replied:
I rewrote this to use `heapq.nlargest` and avoided the need for `key` entirely.

pablogsal commented on src/memray/_memray.pyx:
```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+        for loc, size_and_count in size_and_count_by_location.items()
+    ]
+    top_locations_by_size.sort(key=lambda pair: pair[1], reverse=True)
+    del top_locations_by_size[num_largest:]
```
We may be able to just use `heapq.nlargest` instead of the construction + sort, but I don't know how better or worse in perf would be.

godlygeek replied:
It seems to make very little performance difference one way or the other, but it may scale a bit better in extreme cases, so I've implemented this.

godlygeek commented:
I think I've addressed all of your comments except the note that the docs likely need to be reviewed. I'll check those out on Monday and see what changes are needed.

godlygeek commented:
I've fixed up the docs (only required removing a few sentences) and added a NEWS entry describing the speedups and the fact that it's now easier to interpret the results.

I profiled on a variety of files, and I found that this new implementation was at least 20% faster and up to 50% faster than the one on `main`, even though this implementation looks at all allocations and the one on `main` only looks at high water mark allocations.

godlygeek commented:
I pushed another commit that performs the aggregation in C++ by location rather than by stack, to avoid the need for a second aggregation pass in the Cython code (and to eliminate some ugly code that neither of us really liked).

It has the unfortunate effect of exposing frame IDs from the reader layer to the Cython code and aggregator, though (previously, only stack IDs were exposed).

I used an `unordered_map` for the aggregation - it seems to be around 5% faster to use a vector instead, but that introduces a dependency on even more implementation details (namely, that frame IDs are monotonically increasing numbers that start at 0). I've got the `vector` code stashed if you'd prefer to take the faster but less maintainable version over the slower but more maintainable one.

pablogsal commented:
@godlygeek Seems that you need to rebase :(

pablogsal reviewed (approved):
LGTM
